### ISSUE TITLE
Question about "values"

### ISSUE THREAD

**Issue by jonsterling** [quasar-analytics/sql2-spec#1]
(Putting this question here where it will not be lost in Slack history.)

There’s a subset of the SQL^2 grammar which can be encoded as “precise json”; I’ve been calling this subset the “literals” or the “values”, but it doesn’t seem like there’s any particular such distinction in the implementation of the AST that you showed me. Is that correct? Or is there a precise-json encoding for the entire grammar, beyond what is presented in the Quasar README?

**Comment by jdegoes** [comment]
We need to extract from SQL<sup>2</sup> an Extended JSON (EJSON) subset, which is basically JSON plus characters plus metadata plus whatever literals we've added.

At some point in the future, Quasar will allow reading and writing EJSON instead of the "strict-mode JSON" that is currently supported.

See also [EJSON](https://github.com/ejson).

**Comment by jonsterling** [comment]
@jdegoes sounds cool. Am I right, though, in assuming that regardless of the particular serialization format that we use, there is in fact a subset of the SQL^2 grammar which is of interest for serialization? (e.g. the subset that includes just the literals, etc.)

I'm trying to figure out if it makes sense to arrange the SQL^2 theory as having different syntactic sorts/categories, so that we can conveniently say something like "All terms of sort `VAL` are serializable to (EJSON | Precise JSON | etc)". I've gathered so far that the SQL^2 theory is unisorted, which is possibly nice because it lets you easily compose everything together, but there may be some value in some very course notion of sort that at least allowed us to distinguish between values and queries, etc. But since I don't know the details of the theory, I can't really say whether this is indeed desirable... 

Sorry if these questions are obvious or a nuisance! But without an abstract specification in hand of the language itself in the form of a signature to an algebraic theory (i.e. not just a résumé of concrete notation), it's hard for me to make progress on our PureScript implementation of the syntax, since the answers to these questions will have a great effect on what we do going forward.

**Comment by jdegoes** [comment]
@jonsterling Yep, that's the EJSON grammar! EJSON is a strict superset of JSON but consists purely of literal data. 

Yes, SQL<sup>2</sup> is unisorted, which makes it extremely composable. An EJSON grammar could not easily form a term in the SQL<sup>2</sup> grammar, because the constructors for maps and arrays are not limited to literals, but operate on arbitrary expressions, of which literal is just a special case.

I do think what you want is an EJSON grammar, which would precisely specify the set of all expressions which can be deconstructed to pure literals.

**Comment by jonsterling** [comment]
@jdegoes Aha! WOW, that clarifies so much for me, thanks a ton :smile: 

So, to summarize: Yes, I think what I need is an EJSON grammar. Might be reasonable to construct the algebraic signature (functor) in such a way as it could be composed with the one that contains all the non-literal aspects of SQL^2 in order to get the "real" SQL^2 grammar. Would this suffice?

A few more questions:

1. In SQL^2, I am under the impression that keys in maps (in addition to objects) are arbitrary SQL^2 expressions. Is this correct?

2. What about "SQL^2 Identifiers"? I take it that this is part of SQL^2 proper, not EJSON?

**Comment by jdegoes** [comment]
Sure, happy to help! I think that might work (will be tricky to get right, maybe @sellout has given this some thought).

1. Yes, that's correct. Keys in maps can themselves be, for example, arrays of maps. :smile: 
2. Yes, that's SQL<sup>2</sup>, not EJSON. Identifiers reference aliased datasets, bound variables in a let expression, values imported from maps (whose keys are strings matching the identifiers), and functions in the standard library.

**Comment by jonsterling** [comment]
@jdegoes OK, cool.

1. I think that it should be possible to make this work if the two signatures are something like `EJsonF a` and `QueryF a` (ignore the names!), and then we get `SqlF = EJsonF + QueryF`. The operator for object literals would use `a` for both key and value, and as a result, when we take the least fixed point of this coproduct, we will get a set where the keys can be arrays of maps, etc. Sound reasonable?

2. OK, cool. Is the name of the (e.g.) `DATETIME` function an identifier, for instance?

**Comment by sellout** [comment]
@jonsterling Yeah, exactly right with the `EJSONF :+: QueryF`. The one thing here is that (I think) certain identifiers (like `TIMESTAMP` (is `DATETIME` one?)) need to be included in EJSON, because they are effectively data constructors from literals.

**Comment by jdegoes** [comment]
1. Yes, should work.
2. Yes, that's an identifier.

@sellout I've also been thinking about an alternate way of modeling them, which is as composite types. e.g. `DATE` becomes something with type `{ "month": 'Integer, "year": 'Integer, "day": 'Integer }` (one could use a refinement type as well given our current pattern language for describing types). 

If we went down this path, then it would only be necessary to standardize the composite structural types. Obviously, it would involve some runtime overhead because we'd have to convert from the database's primitive types to the composite ones on-the-fly based on runtime type checks.

But could simplify the model quite a lot and keep EJSON relatively simple, too.

**Comment by jonsterling** [comment]
@sellout OK, cool. (BTW, you're right, it's `TIMESTAMP` not `DATETIME`).

So, one thing I'm not totally clear on, which you've brought up: 

1. In EJSON, presumably `DATETIME,TIME,DATE,INTERVAL` etc. are primitive constructors/operators.
2. In SQL^2, they are currently applications of a free identifier to some data 

Now, in his last comment @jdegoes proposes an interesting approach to simplifying things, but I guess my perspective is that I need to come up with a solution that works *today*, which we can render to the standard Precise JSON format that we use to interact with Quasar. Given those constraints, how do we unify these two conflicting aspects of the grammar?

In my opinion, it would be simplest to go back and say that these are indeed primitive constructors in EJSON, and treat them as such when they are composed into SQL^2—rather than the seemingly-expedient-but-ultimately-IMO-more-complicated approach of treating these as function applications of free identifiers.

What do you two think?

**Comment by jonsterling** [comment]
The other thing is, maybe I've missed part of the idea of identifiers, but it does feel a little weird to see them used to deal with what is an essentially syntactic matter—as if `TIMESTAMP` is the name of an external function or resource that will be supplied, as opposed to part of the grammar of a data object. This seems different from using them to name an object like a dataset, etc.

**Comment by sellout** [comment]
@jonsterling Yeah, that’s my fault … basically, we used to have them as operators that indicated a literal string was to be interpreted as some other type. But now they _are_ functions that can handle arbitrary expressions (that must evaluate to a string to be interpreted as the appropriate type).

**Comment by jonsterling** [comment]
@sellout Oh, interesting... What do these functions evaluate to? Maybe we could have the functions evaluate to whatever structure `EJSON` provides.

**Comment by jdegoes** [comment]
> What do you two think?

Sounds good. Treating them as primitive constructors does not imply they are not implemented via composite types (as described above), and still ensures EJSON is a strict subset of SQL<sup>2</sup>.

**Comment by jonsterling** [comment]
@jdegoes OK, sounds great! For now, I'll implement as primitive constructors w/ strings since that's what Quasar does at the moment, and I'll follow whatever is done later, such as using these composite types (which I assume would be wrapped in these primitive constructors).

**Comment by jonsterling** [comment]
@jdegoes @sellout Currently we encode (something like) EJSON inside JSON itself using this "precise json" format, but this only supports the use of strings as keys for objects.

Is there an alternative encoding that we intend to use? Or is the idea that we will standardize upon a concrete serialization format for EJSON that ignores JSON entirely?

I ask because I'm developing the EJSON library, and since we intend to use it to communicate with Quasar, we need to know what the serialized format will be. For now, one option would be to make serialization fail in case an object has anything but string literals as keys, but it is a bit odd for serialization to be partial.

**Comment by jdegoes** [comment]
@jonsterling Yes, we will use EJSON as the serialization format; and this will be the only way to retrieve data in "strict mode". If you request JSON, you'll get the readable (lossy) encoding.

The only thing that has not been decided is how to represent the top-level set. Most likely, it will be newline-separated, though I'm open to other options, including set literal notation at the top-level.

@sellout ?

**Comment by jonsterling** [comment]
@jdegoes OK, cool. So, for the immediate task, I will provide a lossy encoder into JSON that matches what Quasar is spitting out currently—and in the future we'll build the new order around EJSON first. thanks for the clarification!

**Comment by jdegoes** [comment]
👍
